Optimize DB::combineFilters with BMI2/SSE intrinsics

ZhiguoZh opened the pull request:
As a follow-up of https://github.com/ClickHouse/ClickHouse/pull/56079, the function is optimized with the BMI2 and SSE intrinsics so that devices without AVX512 support could also accelerate the column combinations via data parallelism.
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
The performance experiments of **OnTime** on the Intel server with up to AVX2 (and BMI2) support show that this change could effectively improve the QPS of **Q2** and **Q3** by **5.0%** and **3.7%**  through reducing the cycle ratio of the hotspot, **_DB::MergeTreeRangeReader::ReadResult::optimize_**, **from 11.48% to 1.09%** and **from 8.09% to 0.67%** respectively while having no impact on others.



> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

ZhiguoZh committed:
Enable BMI2 attribute when AVX2 is supported

The BMI2/AVX2 features were introduced with the same architectures
(Haswell from Intel and Excavator from AMD). The enabling of BMI2
attribute should be safe for the AVX2 specific code.

robot-ch-test-poll commented:
<!-- automatic status comment for PR #58800 from ZhiguoZh/ClickHouse:20240115-combine-filter-bmi2 -->
*This is an automated comment for commit ebb3db8787a206a6b34477e447f785cc3f7418c8 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/58800/ebb3db8787a206a6b34477e447f785cc3f7418c8/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/58800/ebb3db8787a206a6b34477e447f785cc3f7418c8/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

alexey-milovidov commented on src/Storages/MergeTree/MergeTreeRangeReader.cpp:
```diff
@@ -1283,6 +1283,72 @@ inline void combineFiltersImpl(UInt8 * first_begin, const UInt8 * first_end, con
+    while (first_begin + bytes_to_read <= first_end)
+    {
+        __m64 src = unalignedLoad<__m64>(first_begin);
+        UInt64 mask = static_cast<UInt64>(_m_to_int64(_mm_cmpeq_pi8(src, __m64 {0})));
```
The usage of MMX alters the mode of operation of x87 FPU:

```
  * Example 1.
  * Compile this code without optimizations:
  *
    #include <vector>
    #include <unordered_set>
    #include <iostream>
    #include <tmmintrin.h>

    int main(int, char **)
    {
        [[maybe_unused]] __m64 shuffled = _mm_shuffle_pi8(__m64{}, __m64{});

        std::vector<int> vec;
        std::unordered_set<int> set(vec.begin(), vec.end());

        std::cerr << set.size() << "\n";
        return 0;
    }

    $ g++ -g -O0 -mssse3 -std=c++17 mmx_bug1.cpp && ./a.out
    terminate called after throwing an instance of 'std::bad_alloc'
    what():  std::bad_alloc

    Also reproduced with clang. But only with libstdc++, not with libc++.

  * Example 2.

    #include <math.h>
    #include <iostream>
    #include <tmmintrin.h>

    int main(int, char **)
    {
        double max_fill = 1;

        std::cerr << (long double)max_fill << "\n";
        [[maybe_unused]] __m64 shuffled = _mm_shuffle_pi8(__m64{}, __m64{});
        std::cerr << (long double)max_fill << "\n";

        return 0;
    }

    $ g++ -g -O0 -mssse3 -std=c++17 mmx_bug2.cpp && ./a.out
    1
    -nan

  * Explanation:
  *
  * https://stackoverflow.com/questions/33692969/assembler-mmx-errors
  * https://software.intel.com/en-us/node/524274
  *
  * Actually it's possible to use 'emms' instruction after decompression routine.
  * But it's more easy to just use 'xmm' registers and avoid using 'mm' registers.
  */
```

ZhiguoZh replied:
Thanks Alexey! I will try to avoid this issue with the suggested method.

alexey-milovidov reviewed (changes requested):
.

ZhiguoZh committed:
Optimize DB::combineFilters with BMI2/SSE intrinsics

The DB::combineFilters combines two UInt8 columns (filters) by
scanning the first column for non-zero elements and replacing them
with elements in the second column contiguously.

As a follow-up of #56079, the function is optimized with the BMI2
and SSE intrinsics so that the x86 devices without AVX512 support
could also accelerate the column combinations via data parallelism.

ZhiguoZh commented:
Hi Alexey, thanks for reviewing the code and for sharing your insights into the MMX mode issue! And performance test on the OnTime workload suggested that the SSE-based implementation could achieve the equivalent performance impact as the MMX version.

@alexey-milovidov

ZhiguoZh commented:
Hi Alexey, I found 5 of the tests failed on the latest version of patch. But it seems most of them are not related to this code change.

![image](https://github.com/ClickHouse/ClickHouse/assets/26588299/5ef0b934-8ba6-44e5-b16d-3c8aa677d04e)

1. The error with the **AST fuzzer (ubsan)** is a known test issue: #57898
2. **Stateful tests (debug)** reported a **timeout** error with **00165_jit_aggregate_functions**
3. **Stateful tests (debug, ParallelReplicas)** reported one **timeout** error with **00082_quantiles** and one "connection reset by peer" error with **00157_cache_dictionary**
4. The reports of two errors in **Performance Comparison Aarch64 [1/4]** cannot be viewed, but this x86-specific change should not have impact on the Aarch64 platform theoretically

As for the 7 regressions in **Performance Comparison [4/4]**, I'm not sure if they are due to the casual instability or if they are critical. Could you share with me some suggestions on their fixes?

@alexey-milovidov

ZhiguoZh commented:
Thanks Alexey!
